packages/block-editor/src/store/selectors.js

```diff
@@ -152,7 +152,9 @@ export const getBlock = createSelector(
 		return {
 			...block,
 			attributes: getBlockAttributes( state, clientId ),
-			innerBlocks: getBlocks( state, clientId ),
+			innerBlocks: areInnerBlocksControlled( state, clientId )
```

Comment:
@youknowriad here is where I'm using the new selector to determine if the blocks should be excluded from the `getBlock` call. I do not notice this call changing anything about edit-site at the moment.

Reply:
you can see that `state.blocks.cache[ clientId ]` (the memoization key a few lines under) is not changing when the flag is set to true or false, which means, the cache key (the object that tells us whether we need to recompute the block object or not, that's a performance technique we're using) is still considering the inner blocks regardless of the flag. This also means a new block object instance is generated regardless of the flag valuue in the state causing an `onChange` call. What we need here is to update the `cache` state to reflect the fact that when a flag is set that particular block and its parents don't need a new cache key when one of their children change. 

I know this cache key technique is a complex logic but it's very important for performance.

Reply:
I haven't figured out how to fix the cache logic yet, but for the sake of investigation, I commented out the dependency so I could see what would happen once I do figure out fixing the cache, and there are a couple further problems:

**Block Editor Equality Check:**

The `BlockEditorProvider` does a [shallow equality check](https://github.com/WordPress/gutenberg/blob/dd9f4544330be4aa32beb96c13a9ad9268c7e573/packages/block-editor/src/components/provider/index.js#L125-L142), so the onChange handler continues firing as you make edits to the template parts (even though template part the InnerBlocks are not part of the given blocks).

If you add a deep equality check there (and "fix" the cache key) then the onChange handler fires only when the top-level template blocks are changed, which fixes the save state.

**Initial blocks**
the `onChange` handler also fires immediately when the editor blocks load, so all of the state is marked as dirty even before you've made any edits.

So I guess we either need to make sure that the object reference only changes when we need it to (is this the purpose of the cache?), or add a deep equality check?

Reply:
So I did some stuff with the caching mechanism. It will now _not_ change the cache key for a block which controls its inner blocks as those inner blocks change. 

But this introduces a new issue. InnerBlocks relies on `getBlock` to get its blocks. If the cache key doesn't change anymore... then getBlock on the parent will not return a new value if its innerblocks change (even if we add logic back for returning innerblocks)

Reply:
Hm. I think I found a (possible?) workaround for that, but I also found an issue with function edits, I think

Reply:
related: #19741